aclk/aclk_otp.c

```diff
@@ -426,112 +268,242 @@ static int aclk_parse_otp_error(const char *json_str) {
     req.port = target->port;
     buffer_sprintf(url, "%s/node/%s/challenge", target->path, agent_id);
-    req.url = url->buffer;
+    req.url = (char *)buffer_tostring(url);
 
     if (aclk_https_request(&req, &resp)) {
         error ("ACLK_OTP Challenge failed");
-        goto cleanup;
+        buffer_free(url);
+        return 1;
     }
     if (resp.http_code != 200) {
         error ("ACLK_OTP Challenge HTTP code not 200 OK (got %d)", resp.http_code);
+        buffer_free(url);
         if (resp.payload_size)
             aclk_parse_otp_error(resp.payload);
         goto cleanup_resp;
     }
-    info ("ACLK_OTP Got Challenge from Cloud");
+    buffer_free(url);
 
-    struct dictionary_singleton challenge = { .key = "challenge", .result = NULL };
+    info ("ACLK_OTP Got Challenge from Cloud");
 
-    if (json_parse(resp.payload, &challenge, json_extract_singleton) != JSON_OK)
-    {
-        freez(challenge.result);
-        error("Could not parse the the challenge");
+    json_object *json = json_tokener_parse(resp.payload);
+    if (!json) {
+        error ("Couldn't parse HTTP GET challenge payload");
         goto cleanup_resp;
     }
-    if (challenge.result == NULL) {
-        error("Could not retrieve challenge JSON key from challenge response");
-        goto cleanup_resp;
+    json_object *challenge_json;
+    if (!json_object_object_get_ex(json, "challenge", &challenge_json)) {
+        error ("No key named \"challenge\" in the returned JSON");
+        goto cleanup_json;
+    }
+    if (!json_object_is_type(challenge_json, json_type_string)) {
+        error ("\"challenge\" is not a string JSON type");
+        goto cleanup_json;
+    }
+    const char *challenge_base64;
+    if (!(challenge_base64 = json_object_get_string(challenge_json))) {
+        error("Failed to extract challenge from JSON object");
+        goto cleanup_json;
+    }
+    if (strlen(challenge_base64) != CHALLENGE_LEN_BASE64) {
+        error("Received Challenge has unexpected length of %zu (expected %d)", strlen(challenge_base64), CHALLENGE_LEN_BASE64);
+        goto cleanup_json;
     }
 
-    // Decrypt the Challenge and Calculate Response
-    size_t challenge_len = strlen(challenge.result);
-    unsigned char decoded[512];
-    size_t decoded_len = base64_decode((unsigned char*)challenge.result, challenge_len, decoded, sizeof(decoded));
-    freez(challenge.result);
-
-    unsigned char plaintext[4096]={};
-    int decrypted_length = private_decrypt(p_key, decoded, decoded_len, plaintext);
-    char encoded[512];
-    size_t encoded_len = base64_encode(plaintext, decrypted_length, encoded, sizeof(encoded));
-    encoded[encoded_len] = 0;
-    debug(D_ACLK, "Encoded len=%zu Decryption len=%d: '%s'", encoded_len, decrypted_length, encoded);
-
-    char response_json[4096]={};
-    sprintf(response_json, "{\"response\":\"%s\"}", encoded);
-    debug(D_ACLK, "Password phase: %s",response_json);
+    *challenge = mallocz(CHALLENGE_LEN);
```

Comment:
Hi Timo! Is it possible that this needs one or two more bytes ? Valgrind has this warning there:

```
==17151== Invalid write of size 1
==17151==    at 0x4A026DB: ??? (in /usr/lib64/libcrypto.so.1.1)
==17151==    by 0x4A02C79: EVP_DecodeUpdate (in /usr/lib64/libcrypto.so.1.1)
==17151==    by 0x26F075: base64_decode_helper (aclk_otp.c:295)
==17151==    by 0x26F075: aclk_get_otp_challenge (aclk_otp.c:380)
==17151==    by 0x26F927: aclk_get_mqtt_otp (aclk_otp.c:474)
==17151==    by 0x269FC9: aclk_attempt_to_connect (aclk.c:657)
==17151==    by 0x269FC9: aclk_main (aclk.c:815)
==17151==    by 0x268458: aclk_starter (aclk_api.c:27)
==17151==    by 0x164FA4: thread_start (threads.c:185)
==17151==    by 0x5069EF6: start_thread (in /lib64/libc.so.6)
==17151==    by 0x50ED54F: clone (in /lib64/libc.so.6)
==17151==  Address 0x170cea51 is 1 bytes after a block of size 256 alloc'd
==17151==    at 0x4840855: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==17151==    by 0x156B07: mallocz (libnetdata.c:176)
==17151==    by 0x26F039: aclk_get_otp_challenge (aclk_otp.c:379)
==17151==    by 0x26F927: aclk_get_mqtt_otp (aclk_otp.c:474)
==17151==    by 0x269FC9: aclk_attempt_to_connect (aclk.c:657)
==17151==    by 0x269FC9: aclk_main (aclk.c:815)
==17151==    by 0x268458: aclk_starter (aclk_api.c:27)
==17151==    by 0x164FA4: thread_start (threads.c:185)
==17151==    by 0x5069EF6: start_thread (in /lib64/libc.so.6)
==17151==    by 0x50ED54F: clone (in /lib64/libc.so.6)
```

I assume might be from `EVP_EncodeUpdate()` that says there needs to be `plus an additional byte for a NUL terminator.` The warning actually appears twice, and from a small check mallocing +2 bytes seems to remove them.

Reply:
Good catch, although the final output is 256 Bytes long it does seem to use extra byte during decoding. This is in line with `EVP_DecodeUpdate` documentation
> For every 4 valid base 64 bytes processed (ignoring whitespace, carriage returns and line feeds), 3 bytes
       of binary output data will be produced (or less at the end of the data where the padding character "=" has been used).

therefore for input 344 long it can use `(344/4)*3` which is 258 bytes during decoding even though final output may be shorter